toolkit/tools/pkg/imagecustomizerlib/customizepackages.go

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package imagecustomizerlib
+
+import (
+	"fmt"
+	"path"
+	"strings"
+
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/imagecustomizerapi"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/logger"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/safechroot"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/shell"
+)
+
+func updatePackages(buildDir string, baseConfigPath string, packagesToAddLists []string, packagesToAdd []string,
+	packagesToRemoveLists []string, packagesToRemove []string, imageChroot *safechroot.Chroot, rpmsSources []string,
+	useBaseImageRpmRepos bool,
+) error {
+	var err error
+
+	allPackagesToRemove, err := collectPackagesList(baseConfigPath, packagesToRemoveLists, packagesToRemove)
+	if err != nil {
+		return err
+	}
+
+	allPackagesToAdd, err := collectPackagesList(baseConfigPath, packagesToAddLists, packagesToAdd)
+	if err != nil {
+		return err
+	}
+
+	err = removePackages(buildDir, allPackagesToRemove, imageChroot)
+	if err != nil {
+		return err
+	}
+
+	err = installPackages(buildDir, allPackagesToAdd, imageChroot, rpmsSources, useBaseImageRpmRepos)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func collectPackagesList(baseConfigPath string, packageLists []string, packages []string) ([]string, error) {
+	var err error
+
+	// Read in the packages from the package list files.
+	var allPackages []string
+	for _, packageListRelativePath := range packageLists {
+		packageListFilePath := path.Join(baseConfigPath, packageListRelativePath)
+
+		var packageList imagecustomizerapi.PackageList
+		err = imagecustomizerapi.UnmarshalYamlFile(packageListFilePath, &packageList)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read package list file (%s):\n%w", packageListFilePath, err)
+		}
+
+		allPackages = append(allPackages, packageList.Packages...)
+	}
+
+	allPackages = append(allPackages, packages...)
+	return allPackages, nil
+}
+
+func removePackages(buildDir string, allPackagesToRemove []string, imageChroot *safechroot.Chroot) error {
+	var err error
+
+	tnfRemoveArgs := []string{
+		"-v", "remove", "--assumeyes", "--disablerepo", "*",
+		// Placeholder for package name.
+		"",
+	}
+
+	// Remove packages.
+	// Do this one at a time, to avoid running out of memory.
+	for _, packageName := range allPackagesToRemove {
+		tnfRemoveArgs[len(tnfRemoveArgs)-1] = packageName
+
+		err = imageChroot.Run(func() error {
+			err := shell.ExecuteLiveWithCallback(tdnfRemoveStdoutFilter, logger.Log.Warn, false, "tdnf",
+				tnfRemoveArgs...)
+			return err
+		})
+		if err != nil {
+			return fmt.Errorf("failed to install package (%s):\n%w", packageName, err)
+		}
+	}
+
+	return nil
+}
+
+// Process the stdout of a `tdnf install -v` call and send the list of installed packages to the debug log.
+func tdnfRemoveStdoutFilter(args ...interface{}) {
```

Comment:
I don't have a strong opinion on this one, but we generally define the callbacks just before we call `ExecuteLiveWithCallback()` in the same function. 

I don't care either way though.

Reply:
Ended up reusing one of the functions for package updates anyway.